### ISSUE TITLE
Frozen Indices

### ISSUE THREAD

**Issue by s1monw** [elastic/elasticsearch#34352]
Frozen indices are intended to allow much higher ratios of disk storage to heap, at the expense of search latency. The idea is to keep frozen indices in an unloaded state (ie not loading lucene datastructures into heap) until they are searched. When a search targets frozen indices, each index would be searched sequentially (instead of in parallel), and each index would be loaded, searched, then unloaded again. Frozen indices would be replicated, unlike closed indices today. In-fact frozen indices are more like open indices with metadata in memory and data fully on disk until it's needed to be loaded

Frozen indices will be available in the default distribution but not in the pure OSS distribution.

In oder to implement this, the feature is broken down to the following steps:

 - [x] Add the ability to force searches to go through a threadpool that has a limited amount of threads to ensure shards data is materialized sequentially or with very limited concurrency. This will allow frozen indices to be marked as `index.search.throttled` that will force searches as well as other data access like `_get` or `_explain` through a dedicated `search_throttled` threadpool which has only 1 thread by default. (#33732)
 - [ ] Prevent throttled indices to be excluded in `_search` and `_msearch` by default. This will prevent frozen indices to be included in wildcard or alias searches by default but allows opt-in to searching these indices. At the same time these indices are treated like ordinary indices in all other APIs
 - [ ] Add a _frozen_ engine that maintains static stats and opens an index reader on demand. The searcher / reader returned form this engine will be able to reset itself after it search phase to release resources and open up a new searcher / reader from the same commit point on subsequent search phases like fetch or scan / scroll
 - [ ] Add an API that allows to open a closed index as frozen. This is the simplest way to introduce frozen indices, further improvements on how we go from Open to Frozen can be done as followups if necessary. As a start an index must got through Closed in order to be frozen or unfrozen.
 - [ ] Add an optimization that returns a special index reader that allows range queries on date fields to rewrite in `can_match` phase in order to exclude frozen indices in `can_match` without opening up their index reader.

**Comment by s1monw** [comment]
@debadair where do you see the documentation go for this feature?

**Comment by s1monw** [comment]
I thought about some future work on this and wanted to put it here for awareness and potential discussion:

- The current design would open a reader before we check the cache for hits and even if we cache we would invalidate the cache after each search phase. I wonder if we can trick the lazy reader design into having it's own cache key helper that we can trick into surviving close calls of the actual reader. That way we might be able to allow request cache hits on frozen indices without actually opening the reader

 - should we allow forcemerge on a frozen index? This might allow us to freeze first and the select a new codec to get best compression. It’s relatively straight forward to do with our infrastructure and an IW that is not pulling NRT readers. And we have limited threads doing this anyway in the force merge threadpool. That's totally something we can do way later.

**Comment by cdahlqvist** [comment]
When benchmarking frozen indices to get an idea of the overhead and added latency it adds I can see two common scenarios:

**1) Dedicated Cold nodes**

Here all indices will be frozen, and due to the low heap usage it may be possible to hold a lot of data with a small heap. In order to efficiently use the available resources and allow parallel processing of frozen indices, I assume it may make sense to set the number of threads used for frozen indices processing to the number of CPU cores on the host. This naturally assumes there is enough heap available to support this.

**2) Warm nodes**

Frozen indices could also be used to make Warm nodes able to handle larger amounts of data. Here the most recent and generally most queried data would be open, while older indices that are not queried frequently could be frozen. The ratio of frozen to open indices will naturally vary from use-case to use-case, but for the benchmark a 50% ratio might be a good starting point. In this scenario, a single thread used to handle frozen indices is probably appropriate.

_Does this make sense?_

In addition to benchmarking these two scenarios, I would also like to benchmark using indices that have and have not been forcemerged down to a single segment since restoration time here may vary. If we add a reference benchmark for each scenario with all indices open, we get a total of 8 benchmark runs.

**Benchmark set-up**

Test data would be generated using the rally-eventdata-track. A single index with 2 primary shards and around 50GB in size would probably be reasonably realistic. Snapshots would be created before and after forcemerging it down to a single segment. This index can then be restored and renamed multiple times to reach a reasonable data volume (~2TB?). Querying would simulate the standard Kibana dashboards available in the rally-eventdata-track. As Kibana dashboards can require a good amount of processing, I would probably run querying in a single thread.

As querying using Rally is quite light, we should be able to run this on a single host. An **AWS EC2 d2.xlarge** instance might be a good choice as it has enough disk space and used the type of slower storage we often see for these node types.

**Comment by s1monw** [comment]
> Here all indices will be frozen, and due to the low heap usage it may be possible to hold a lot of data with a small heap. In order to efficiently use the available resources and allow parallel processing of frozen indices, I assume it may make sense to set the number of threads used for frozen indices processing to the number of CPU cores on the host. This naturally assumes there is enough heap available to support this.

I would like to keep the thread-settings for this to the default. We can still test this but that is for a later iteration.

> Frozen indices could also be used to make Warm nodes able to handle larger amounts of data. Here the most recent and generally most queried data would be open, while older indices that are not queried frequently could be frozen. The ratio of frozen to open indices will naturally vary from use-case to use-case, but for the benchmark a 50% ratio might be a good starting point. In this scenario, a single thread used to handle frozen indices is probably appropriate.

In general I think this scenario is rare and benchmarking will be tricky. I don't want to give the impression that this is a recommended setup so I'd rather not advertise it too much.

**Comment by cdahlqvist** [comment]
I have run a first set of benchmarks against frozen indices. I based these runs on [this rally-eventdata-challenge](https://github.com/cdahlqvist/rally-eventdata-track/blob/cd-frozen/eventdata/challenges/frozen.json). As some of the queries are expected to take quite long I only ran 20 iterations, and as we are simulating cold nodes I did not include any warm-up iterations, so all requests are included in the results.

This benchmark was run against a single **AWS EC2 d2.xlarge node** with disks arranged in RAID0. 20 indices of two ~25GB shards were uploaded (total of ~1TB) and the simulated Kibana queries ran against all of these. I did set `pre_filter_shard_size` to a high value (100000) to get around an open issue, which means the recommended initial match check was not run.

The first set of benchmarks were run against indices that had been forcemerged down to a single segment. The simulated **Discover application** results ina single query/aggregation while the simulated **Content Issues dashboard** consists of 5 visualizations that result in 5 aggregation queries. I ran 3 series for varying number of threads in the `search_throttled` thread pool.

Even though the sample size is small I have included 50th and 100th percentile values from the Rally runs. 

**Results querying Discover application**

| Frozen | Throttled Search Threads |  50th Percentile |  100th Percentile  |
|--------|:-------------:|:------:|:-----:|
| No | 1 | 6.452s | 21.905s |
| Yes | 1 | 9.968s | 32.265s |
| No | 2 | 6.270s | 113.830s |
| Yes | 2 | 7.442s | 37.752s |
| No | 4 | 6.139s | 80.739s |
| Yes | 4 | 5.871s | 37.770s |

**Results querying Content Issues dashboard**

| Frozen | Throttled Search Threads |  50th Percentile |  100th Percentile  |
|--------|:-------------:|:------:|:-----:|
| No | 1 | 4.650s | 34.364s |
| Yes | 1 | 50.104s | 98.768s |
| No | 2 | 3.504s | 8.863s |
| Yes | 2 | 47.298s | 70.853s |
| No | 4 | 3.140s | 8.854s |
| Yes | 4 | 40.845s | 100.777s |

**Querying non-forcemerged indices** 

After this had completed I restored 20 indices that had not been forcemerged at all. 

When running the benchmark against these in open state I got the following latencies:

| Query |  50th Percentile |  100th Percentile  |
|--------|:-------------:|:------:|
| Discover | 16.186ss | 57.659s |
| Content Issues | 7.737s | 201.562s |

When trying to run this on frozen indices all requests however hit the 10 minute time-out I had configured, and that part of the benchmark failed.

Given this, might it perhaps be appropriate to require indices to be forcemerged down to a single segment before being frozen or at least issue a warning?

**Comment by s1monw** [comment]
This issue is basically done. I am keeping this open in order to publish numbers from benchmarks in the near future.

**Comment by cdahlqvist** [comment]
I have now run a first set of benchmarks for frozen indices where the number of indices in the cluster gradually increases from 1 to 18. Each index consists of 2 shards and is ~50GB in size and was generated and queried using the rally-eventdata-track. This was run against a build with the build hash `447e5d2`.

Querying was done using two types of simulated Kibana queries:

- **Kibana Discover query:** Single histogram aggregation and 500 documents being retrieved
- **Content Issues Dashboard query:** Small dashboard containing 5 aggregations, aggregating over a subset of the data.

For indices forcemerged down to a single segment, two different heap sizes were used when querying frozen indices: 1GB and 15GB. For the larger heap, a reference run with all indices open was performed at every step. Each run consists of 20 queries with a varying time interval, and 50% and 100% percentile service time values are shown.

The results are shown below in the form of Kibana screenshots from the Rally metrics store.

**Discover query, 15GB Heap, Forcemerged indices, all indices open and not frozen**

![screenshot 2018-12-02 at 15 46 42](https://user-images.githubusercontent.com/2804455/49643929-fdc1db00-f9e4-11e8-8857-a29451ca9688.png)

**Discover query, 1GB/15GB Heap, Forcemerged indices, all indices frozen**

![screenshot 2018-12-02 at 15 47 25](https://user-images.githubusercontent.com/2804455/49643941-061a1600-f9e5-11e8-82c7-105e7b6c722e.png)

**Content Issues Dashboard query, 15GB Heap, Forcemerged indices, all indices open and not frozen**

![screenshot 2018-12-02 at 15 47 42](https://user-images.githubusercontent.com/2804455/49643951-0d412400-f9e5-11e8-8d4b-f350e3f8d640.png)

**Content Issues Dashboard query, 1GB/15GB Heap, Forcemerged indices, all indices frozen**

![screenshot 2018-12-02 at 15 48 01](https://user-images.githubusercontent.com/2804455/49643963-129e6e80-f9e5-11e8-8c0b-a24e18651025.png)

**Comment by kpstolk** [comment]
> should we allow forcemerge on a frozen index?

@s1monw Maybe I missed the answer in this thread, but this is/won't be allowed?
Because the docs mention that a `_forcemerge` should only be called on read-only indices it made sense to me to use `_freeze` because it does exactly this: make it read-only.

But it took me a while to figure out that `_forcemerge` doesn't do anything on a frozen index and doesn't fail either..
